### ISSUE TITLE
CustomPickler fallback on default generated pickler

### ISSUE THREAD

**Issue by Bananas-Are-Yellow** [nessos/FsPickler#29]
I want to serialize a _subset_ of a graph. When I deserialize later, I will reconnect with the external objects. To do that, I use my own tag-to-object table. When I first encounter an external object, I can write out the tag followed by the object. If I encounter it again, I can just write out the tag. I can use the `StreamingContext` to pass this state through the pickling, so that part is fine.

I can use the `[<CustomPickler>]` attribute and supply a pickler that does all this, but having written the tag, I then have to manually pickle the object. What I'd like to do is have some way for the custom pickler to obtain the pickler that would have been generated (as if the `[<CustomPickler>]` attribute had not been applied), so that I can use that to write out the object. Likewise when reading.

Is there some way to do this? The only way I could think of is for all my references to use a `Wrapper<'T>` struct, which holds the reference, so that the wrapper can have the custom pickler, but the referenced type `'T` inside it does not, but I don't really want to complicate all my types and code just for this.

Thanks,

David.

**Comment by eiriktsarpalis** [comment]
Reading again this issue, it seems clear to that your initial requirement can be solved without having to change pickler generation logic. This can be possibly addressed by adding support for serialising object graphs using holes. For example:
```fsharp
FsPickler.SerializeWithHoles<Foo>(stream, foo, [x : obj ; y : obj ; z : obj])
```
This would serialise the graph, annotating any occurrence of x, y or z as a hole in the serialization. Its dual method
```fsharp
FsPickler.DeserializeWithHoles<Foo>(stream, [x' : obj ; y' : obj ; z' : obj])
```
would deserialize, patching any hole occurrence with the object of indicated index. Note that this cannot be type-safe and will not work well with cyclic graphs. Importantly, this addition is unrelated to picklers per se, but rather pertains to augmenting the graph traversal logic.

**Comment by Bananas-Are-Yellow** [comment]
In the end, I wrote my own pickler library. I hope you are not offended. It took me 2 weeks: one week to write the pickling logic itself, and another week to figure out the code generation. It was a good exercise and I learned a lot from doing it.

My requirements are different from yours.

I am focused on data storage, rather than transmitting data. This requires persistent links to external objects already stored, as we have discussed. The caller creates a pickler table, providing a function to override the default pickler. Typically this function will wrap the default pickler with an outer pickler that makes some decisions before calling the default pickler. When pickling is performed, the write state provides access to a caller-supplied *context*, which can store tables used by this outer pickler.

I am only interested in binary format, because I need the fastest and most compact data storage. I have implemented a text format for writing only, which is useful for visualizing and debugging the output. I don't need JSON for data storage, although I use Json.NET for data transmission over web sockets, which is a separate part of my application.

My data is a tree structure and I store child nodes in arrays. I generally use arrays a lot, since they are a compact representation. I wanted to avoid writing out an `ObjectFlags` field before every item in the array, even if it is only a single byte. If the type is not sealed, I generate a polymorphic pickler, which wraps the pickler for the type and records the actual type.

I am only interested in F#, so I have no interest in reference cycles or closures or legacy serialization methods. If I am using an existing type, which has internal cycles (e.g. `Dictionary<'k,'v>`), I'd rather have that serialized in a compact form.

The hardest part was the code generation. I didn't want to teach myself incomprehensible IL and Reflection Emit. So I use F# quotations and compile them using `FSharp.Quotations.Evaluator`. However, this has the limitation that it does not support `Expr.FieldSet` expressions, so I fall back on LINQ expressions in that area.

Lastly, my `Pickler<'t>` is a record, like your documentation, but unlike your code. I guess that's how you started out.
